syntax = "proto3";

package Chat;

import "google/protobuf/timestamp.proto";

enum ContactsAction
{
	CONTACTS_ACTION_UNSPECIFIED = 0;
	CONTACTS_ACTION_ACCEPT = 1;
	CONTACTS_ACTION_REJECT = 2;
	CONTACTS_ACTION_REQUEST = 3;
	CONTACTS_ACTION_REMOVE = 4;
}

enum ContactStatus
{
	CONTACT_STATUS_UNSPECIFIED = 0;
	CONTACT_STATUS_ACCEPTED = 1;
	CONTACT_STATUS_INCOMING = 2;
	CONTACT_STATUS_OUTGOING = 3;
	CONTACT_STATUS_REJECTED = 4;
}

enum UserStatus
{
	USER_STATUS_UNSPECIFIED = 0;
	USER_STATUS_ONLINE = 1;
	USER_STATUS_OFFLINE = 2;
}

enum ContactsActionServer
{
	CONTACTS_ACTION_SERVER_UNSPECIFIED = 0;
	CONTACTS_ACTION_SERVER_ADD = 1;
	CONTACTS_ACTION_SERVER_REMOVE = 2;
	CONTACTS_ACTION_SERVER_UPDATE = 3;
};

enum OtcResult
{
	OTC_RESULT_UNSPECIFIED = 0;
	OTC_RESULT_ACCEPTED = 1;
	OTC_RESULT_REJECTED = 2;
	OTC_RESULT_CANCELED = 3;
	OTC_RESULT_EXPIRED = 4;
}

// Keep in sync with bs::network::ChatOTCSide::Type
enum OtcSide
{
	OTC_SIDE_UNDEFINED = 0;
	OTC_SIDE_BUY = 1;
	OTC_SIDE_SELL = 2;
};

// Keep in sync with bs::network::OTCRangeID::Type
enum OtcRangeType
{
	OTC_RANGE_1_5     = 0;
	OTC_RANGE_5_10    = 1;
	OTC_RANGE_10_50   = 2;
	OTC_RANGE_50_100  = 3;
	OTC_RANGE_100_250 = 4;
	OTC_RANGE_250PLUS = 5;
};

message OtcRange
{
	uint64 lower = 1;
	uint64 upper = 2;
}

message Data
{
	enum Direction
	{
		NOT_SET = 0;
		SENT = 1;
		RECEIVED = 2;
	};

	message Room
	{
		string id = 1;
		string owner_id = 2;
		string title = 3;
		string key = 4;
		bool is_private = 5;
		bool disable_send_user_updates = 6;
		bool disable_display_user_list = 7;
		bool disable_display_tray_notification = 8;
		bool is_trading_available = 9;
	}

	message ContactRecord
	{
		string user_id = 1;
		string display_name = 2;
		string contact_id = 3;
		ContactStatus status = 4;
		bytes public_key = 5;
	}

	message User
	{
		string user_id = 1;
		UserStatus status = 2;
	}

	message Message
	{
		enum State
		{
			UNDEFINED                 = 0x0000;
			INVALID                   = 0x0001;
			ACKNOWLEDGED              = 0x0002;
			READ                      = 0x0004;
			SENT                      = 0x0008;
		};

		enum Encryption {
			UNENCRYPTED = 0;
			IES = 1;
			AEAD = 2;
		};

		message OtcRequest
		{
			OtcSide side = 1;
			OtcRangeType range_type = 2;
		}

		message OtcResponse
		{
			OtcSide side = 1;
			OtcRange price = 2;
			OtcRange quantity = 3;
		}

		message OtcUpdate
		{
			uint64 amount = 1;
			uint64 price = 2;
		}

		message OtcCloseTrading
		{
		}

		string id = 1;
		google.protobuf.Timestamp date_time = 2;
		string sender_id = 3;
		string receiver_id = 4;
		uint32 state = 5;
		Encryption encryption = 6;
		bytes nonce = 7;
		string message_payload = 8;

		// FIXME: remove this internal field (security problem)
		bool loaded_from_history = 9;

		string message_text = 10;

		oneof otc
		{
			OtcRequest request = 20;
			OtcResponse response = 21;
			OtcUpdate update = 22;
			OtcCloseTrading close_trading = 23;
		}
	}

	message OtcCloseTrading
	{
	}

	oneof data
	{
		Room room = 1;
		ContactRecord contact_record = 2;
		User user = 3;
		Message message = 4;
	}

	Direction direction = 20;
}


message Request
{
	message Login
	{
	}

	message Logout
	{
	}

	message SendMessage
	{
	}

	message Messages
	{
	}

	message OnlineUsers
	{
	}

	message AskForPublicKey
	{
	}

	message SendOwnPublicKey
	{
	}

	message ChangeMessageStatus
	{
	}

	message ContactsActionDirect
	{
	}

	message ContactsActionServer
	{
	}

	message ChatroomsList
	{
	}

	message SendRoomMessage
	{
	}

	message ContactsList
	{
	}

	message SearchUsers
	{
	}

	message SessionPublicKey
	{
	}

	message ReplySessionPublicKey
	{
	}

	oneof data
	{
		Login login                                             = 1;
		Logout logout                                           = 2;
		SendMessage send_message                                = 3;
		Messages messages                                       = 4;
		OnlineUsers online_users                                = 5;
		AskForPublicKey ask_for_public_key                      = 6;
		SendOwnPublicKey send_own_public_key                    = 7;
		ChangeMessageStatus change_message_status               = 8;
		ContactsActionDirect contacts_action_direct             = 9;
		ContactsActionServer contacts_action_server             = 10;
		ChatroomsList chatrooms_list                            = 11;
		SendRoomMessage send_room_message                       = 12;
		ContactsList contacts_list                              = 13;
		SearchUsers search_users                                = 14;
		SessionPublicKey session_public_key                     = 15;
		ReplySessionPublicKey reply_session_public_key          = 16;
	}
}

message Response
{
	message Login
	{
	}

	message Messages
	{
	}

	message Success
	{
	}

	message Error
	{
	}

	message UsersList
	{
	}

	message AskForPublicKey
	{
	}

	message SendOwnPublicKey
	{
	}

	message PendingMessage
	{
	}

	message SendMessage
	{
	}

	message ChangeMessageStatus
	{
	}

	message ContactsActionDirect
	{
	}

	message ContactsActionServer
	{
	}

	message ChatroomsList
	{
	}

	message RoomMessages
	{
	}

	message ContactsList
	{
	}

	message SearchUsers
	{
	}

	message Logout
	{
	}

	message SendSessionPublicKey
	{
	}

	message SessionPublicKey
	{
	}

	message ReplySessionPublicKey
	{
	}

	oneof data
	{
		Login login                                             = 1;
		Messages messages                                       = 2;
		Success success                                         = 3;
		Error error                                             = 4;
		UsersList users_list                                    = 5;
		AskForPublicKey ask_for_public_key                      = 6;
		SendOwnPublicKey send_own_public_key                    = 7;
		PendingMessage pending_message                          = 8;
		SendMessage send_message                                = 9;
		ChangeMessageStatus change_message_status               = 10;
		ContactsActionDirect contacts_action_direct             = 11;
		ContactsActionServer contacts_action_server             = 12;
		ChatroomsList chatrooms_list                            = 13;
		RoomMessages room_messages                              = 14;
		ContactsList contacts_list                              = 15;
		SearchUsers search_users                                = 16;
		Logout logout                                           = 17;
		SendSessionPublicKey send_session_public_key            = 18;
		SessionPublicKey session_public_key                     = 19;
		ReplySessionPublicKey reply_session_public_key          = 20;
	}
}
